lib/logic/player/content.dart

```diff
@@ -756,17 +836,30 @@ class ContentUtils {
            queues.type != QueueType.origin && origin == PlaybackControl.instance.currentSongOrigin;
   }
 
-  /// Returns a default icon for a [PersistentQueue].
-  static IconData persistentQueueIcon(PersistentQueue queue) {
-    return persistentQueuePick<PersistentQueue, IconData>(
-      contentType: queue.runtimeType,
-      album: Album.icon,
+  /// Returns a default icon for a [Content].
+  static IconData contentIcon<T extends Content>([Type? contentType]) {
+    return contentPick<T, IconData>(
+      contentType: contentType,
+      song: Song.icon,
+      album: Album.icon, 
       playlist: Playlist.icon,
+      artist: Artist.icon,
+    );
+  }
+
+  /// Returns a string which represents a given content type.
+  static String contentTypeString<T extends Content>([Type? contentType]) {
```

Comment:
Renamed to `contentTypeId`

To be honest, I didn't understand what you meant with the second part of your comment. Can you elaborate?

Reply:
> Can you elaborate?

We currently have the following class hierarchy:
```text
~Content
- *Song
- Genre
- ~SongOrigin
  - *Artist
  - ~PersistentQueue
    - *Album
    - *Playlist
```
I marked abstract classes with a `~` and classes and classes returned by `Content.enumerate` with a `*`.

At first I thought that the `Content` class would have fields and functionality which were to generic and wouldn't apply to some of it's child classes, for example I thought `contentIcon` might be there to return the icon of the content of a list of contents and would therefore be better moved to `SongOrigin`. I then realized that we are actually trying to access static variables via a dynamic type, e.g. `contentIcon` resolves to a static `icon` field of one of the content subclasses.

I also saw that we accepted `Content` at some places in the code where we actually just wanted to accept one of the types returned by `Content.enumerate`. I saw that those were not all non-abstract child classes of `Content` and thought that these classes then either shouldn't inherit from `Content` or that there should be a mixin class to allow accepting only these classes that are returned by `Content.enumerate`. I realized that it's just the `Genre`, which seems like it's just a work in progress right now.

So there are two problems here:
1. We want to have static data shared between all instances of a non-abstract content type and be able to access it dynamically.
2. We want to ensure that all content types are handled.

The Dart language unfortunately doesn't allow for either abstract statics or resolving statics via `Type`, so we currently have `contentPick` to help with that, where we specify each static manually. This approach has a drawback that we have to manually resolve the different statics. We want to do `contentType.icon`, but we have to do
```dart
contentPick<T, IconData>(
  contentType: contentType,
  song: Song.icon,
  album: Album.icon, 
  playlist: Playlist.icon,
  artist: Artist.icon,
);
```
This also means that the separation of concerns is broken in most usages of `contentPick`, because specific knowledge about constants and variables of each content type is required. When we add a new content type like the `Genre` and we include it in `contentPick`, there will be errors all over the codebase, instead of mostly in the `Genre` class, where they should be. This of course doesn't apply to all usages of `contentPick`.

We also currently need `contentPick` because `Content.enumerate` returns a list of `Type` and the compiler can't help us to verify that an `if` or `switch` would be exhaustive.


To solve this, I propose that we add a `ContentType` enum, which holds all static information about a content type. This requires the new enhanced enums (and therefore `sdk: '>=2.17.0 <3.0.0'`). `Content` will gain an abstract `ContentType get type;` getter, which forces every content subclass to have a content type and allows for dynamic access of the static info. This way we can force each `ContentType` to have the required information and can just write `content.type.icon`. Furthermore, because the compiler knows all possible values of `ContentType`, when writing a function that returns a non nullable value based on a `ContentType`, the code won't compile unless all content types are handled. For switch statements, this can even be improved generally by enabling `exhaustive_cases` and setting it's severity to `error`. 


<details>
<summary>Here is a demonstration on how the implementation could look like.</summary>

```dart
enum ContentType {
  song(1),
  artist(2);

  const ContentType(this.icon);
  final int icon;
}

abstract class Content {
  const Content();
  ContentType get type;
  String get id;
}

class Song extends Content {
  const Song(this.id);
  
  @override
  ContentType get type => ContentType.song;

  @override
  final String id;
}

class Artist extends Content {
  const Artist(this.id);
  
  @override
  ContentType get type => ContentType.artist;

  @override
  final String id;
}

class ContentTuple {
  final List<Song> songs;
  final List<Artist> artists;
  const ContentTuple({this.songs = const [], this.artists = const []});

  List<T> get<T extends Content>(ContentType type) {
    switch(type) {
      case ContentType.song:
        return songs as List<T>;
      case ContentType.artist:
        return artists as List<T>;
    }
  }

  List<Content> get merged => [for (final contentType in ContentType.values) ...get(contentType)];
}

extension RequireMap<K, V> on Map<K, V> {
  V requireValue(K key) {
    final value = this[key];
    if (value != null) {
      return value;
    }
    if (value is V && containsKey(key)) {
      return value;
    }
    throw ArgumentError('No entry for key $key');
  }
}

class ContentMap<V> {
  const ContentMap(this.songValue, this.artistValue);
  factory ContentMap.from(Map<ContentType, V> map) {
    assert(map.length == ContentType.values.length);
    return ContentMap(map.requireValue(ContentType.song), map.requireValue(ContentType.artist));
  }
  
  final V songValue;
  final V artistValue;
  
  V get(ContentType type) {
    switch(type) {
      case ContentType.song:
        return songValue;
      case ContentType.artist:
        return artistValue;
    }
  }
  
  List<V> get values => [for (final type in ContentType.values) get(type)];
}
```
Notice that the compiler won't allow compilation of the `ContentTuple` and `ContentMap` when adding a new value because of their respective `get` functions.

Example usage:
```dart
final serializerPaths = ContentMap<String>.from({
  for (final contentType in ContentType.values) contentType: 'favorites_${contentType.name}.json'
});
print(serializerPaths.values);
const ContentMap<int?> nullableContentMap = ContentMap(1, null);
print(nullableContentMap.get(ContentType.song)!);
print(nullableContentMap.get(ContentType.artist));
const c = ContentTuple(songs: [Song('song1')]);
print(c.get<Song>(ContentType.song));
print(c.merged.any((content) => content.id == 'song1'));
```
</details>

Reply:
> I marked abstract classes with a `~` and classes and classes returned by `Content.enumerate` with a `*`.

Yes, we can think of `*` as of leafs in this hierarchy.

> At first I thought that the `Content` class would have fields and functionality which were to generic and wouldn't apply to some of it's child classes,.

You are right that `Content` is there to declare common properties of all the content classes, as well as to allow to work with content generically.

>  for example I thought `contentIcon` might be there to return the icon of the content of a list of contents and would therefore be better moved to `SongOrigin`.

There is no use case to have an icon for Content or SongOrigin, only for specific types of content, which, more broadly, applies to `contentPick` as well - it generally is only supposed to handle leaf classes. There's `fallback`, but you probably can already see why it's called so.

> I saw that those were not all non-abstract child classes of `Content` and thought that these classes then either shouldn't inherit from `Content` or that there should be a mixin class to allow accepting only these classes that are returned by `Content.enumerate`.

What problem does this solve? What do you do for cases where you need to be able to handle both `Content` and its leafs such as ContentSelectionController and other selection actions?

Regarding your proposed enum solution - what will you do when you need to make a class generic and pass a type to, let's say `ContentSelectionController<T>` or `PersistentQueueTile<T>` type argument? It would be nice to have a static analysis for all this code, but `contentPick` is as much as we can achieve here, I believe.

There are just too many questions here and I still don't understand the essense of the problem you are trying to solve -both of the problems you mentioned are already solved by `contentPick`.

Also, if you want to continue this discussion, please open an issue. I'm definitely not refactoring the enire app here.